File: app/src/main/java/com/druk/servicebrowser/RegistrationManager.java
package com.druk.servicebrowser;

import android.content.Context;
import android.net.nsd.NsdManager;
import android.net.nsd.NsdServiceInfo;
import android.util.Log;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RegistrationManager {

    private static final String TAG = "RegistrationManager";

    private final NsdManager nsdManager;
    private final Map<BonjourServiceInfo, NsdManager.RegistrationListener> mRegistrations = new HashMap<>();

    public RegistrationManager(Context context) {
        nsdManager = BonjourApplication.getNsdManager(context);
    }

    public void register(BonjourServiceInfo bonjourServiceInfo, RegistrationCallback callback) {
        NsdServiceInfo nsdServiceInfo = new NsdServiceInfo();
        nsdServiceInfo.setServiceName(bonjourServiceInfo.getServiceName());
        nsdServiceInfo.setServiceType(bonjourServiceInfo.getRegType());
        nsdServiceInfo.setPort(bonjourServiceInfo.getPort());

        for (Map.Entry<String, String> entry : bonjourServiceInfo.getTxtRecords().entrySet()) {
            nsdServiceInfo.setAttribute(entry.getKey(), entry.getValue());
        }

        NsdManager.RegistrationListener listener = new NsdManager.RegistrationListener() {
            @Override
            public void onServiceRegistered(NsdServiceInfo serviceInfo) {
                BonjourServiceInfo registered = new BonjourServiceInfo.Builder()
                        .serviceName(serviceInfo.getServiceName())
                        .regType(bonjourServiceInfo.getRegType())
                        .port(bonjourServiceInfo.getPort())
                        .txtRecords(bonjourServiceInfo.getTxtRecords())
                        .build();
                mRegistrations.put(registered, this);
                callback.onServiceRegistered(registered);
            }

            @Override
            public void onRegistrationFailed(NsdServiceInfo serviceInfo, int errorCode) {
                Log.e(TAG, "Registration failed: " + errorCode);
                callback.onRegistrationFailed(errorCode);
            }

            @Override
            public void onServiceUnregistered(NsdServiceInfo serviceInfo) {
                Log.i(TAG, "Service unregistered: " + serviceInfo.getServiceName());
            }

            @Override
            public void onUnregistrationFailed(NsdServiceInfo serviceInfo, int errorCode) {
                Log.e(TAG, "Unregistration failed: " + errorCode);
            }
        };

        nsdManager.registerService(nsdServiceInfo, NsdManager.PROTOCOL_DNS_SD, listener);
    }

    public void unregister(BonjourServiceInfo service) {
        NsdManager.RegistrationListener listener = mRegistrations.remove(service);
        if (listener != null) {
            nsdManager.unregisterService(listener);
        }
    }

    public List<BonjourServiceInfo> getRegisteredServices() {
        return Collections.unmodifiableList(new ArrayList<>(mRegistrations.keySet()));
    }

    public interface RegistrationCallback {
        void onServiceRegistered(BonjourServiceInfo service);
        void onRegistrationFailed(int errorCode);
    }
}

File: app/src/main/java/com/druk/servicebrowser/RegistrationManager.kt
package com.druk.servicebrowser

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.util.Log
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class RegistrationManager(context: Context) {

    private val nsdManager = BonjourApplication.getNsdManager(context)
    private val registrations = HashMap<BonjourServiceInfo, NsdManager.RegistrationListener>()

    suspend fun register(bonjourServiceInfo: BonjourServiceInfo): BonjourServiceInfo =
        suspendCancellableCoroutine { cont ->
            val nsdServiceInfo = NsdServiceInfo().apply {
                serviceName = bonjourServiceInfo.displayName
                serviceType = bonjourServiceInfo.regType
                port = bonjourServiceInfo.port
                bonjourServiceInfo.txtRecords.forEach { (key, value) ->
                    setAttribute(key, value)
                }
            }

            val listener = object : NsdManager.RegistrationListener {
                override fun onServiceRegistered(serviceInfo: NsdServiceInfo) {
                    val registered = BonjourServiceInfo(
                        serviceName = serviceInfo.serviceName,
                        regType = bonjourServiceInfo.regType,
                        port = bonjourServiceInfo.port,
                        txtRecords = bonjourServiceInfo.txtRecords
                    )
                    registrations[registered] = this
                    cont.resume(registered)
                }

                override fun onRegistrationFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
                    Log.e(TAG, "Registration failed: $errorCode")
                    cont.resumeWithException(RegistrationException(errorCode))
                }

                override fun onServiceUnregistered(serviceInfo: NsdServiceInfo) {
                    Log.i(TAG, "Service unregistered: ${serviceInfo.serviceName}")
                }

                override fun onUnregistrationFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
                    Log.e(TAG, "Unregistration failed: $errorCode")
                }
            }

            cont.invokeOnCancellation {
                try {
                    nsdManager.unregisterService(listener)
                } catch (_: IllegalArgumentException) {
                }
            }

            nsdManager.registerService(nsdServiceInfo, NsdManager.PROTOCOL_DNS_SD, listener)
        }

    fun unregister(service: BonjourServiceInfo) {
        registrations.remove(service)?.let { listener ->
            nsdManager.unregisterService(listener)
        }
    }

    fun getRegisteredServices(): List<BonjourServiceInfo> =
        registrations.keys.toList()

    class RegistrationException(val errorCode: Int) : Exception("Registration failed: $errorCode")

    companion object {
        private const val TAG = "RegistrationManager"
    }
}
